### PR TITLE
feat: Fail on load-errors.

### LINKED ISSUE #855: Throw an error on not found browsers
Currently, it looks like if a browser isn't able to be loaded, karma just warns you and happily proceeds with the browsers that could be loaded.

https://github.com/karma-runner/karma/blob/master/lib/launcher.js#L41-L46

In a CI environment, it would be really, really easy to overlook that all of the sudden one of your configured browsers isn't working anymore, but since no errors are thrown the build is still happy.

I'm proposing that instead of continuing on, encountering a browser that couldn't be loaded should be treated as a fatal error - any loaded browsers should be cleaned up, and the process should exit.

### PR DESCRIPTION
**Pull request by budde377**
Stop karma if a launcher, reporter, plugin, or preprocessor fails to
load, after attempting to load each of them, and reporting errors for
each load error. 

~~Also someone should prob. tell @nikku that the karma-browserify 
plugin is going to stop working due the changed signature of 
`createPreprocessor`.~~

Closes #855

### CODE REVIEW

**Comment by budde377** [comment]
@srawlins, can you sign the CLA?

**Comment by nikku** [comment]
@nikku does not like that change for the reason you mentioned (`createPreprocessor` syntax changed). Isn't there another way to implement that in a backward compatible manner?

**Comment by budde377** [comment]
@nikku Sorry for the indirection. 
I might have an idea.

**Comment by dignifiedquire** [comment]
@budde377 is this ready for review?

**Comment by budde377** [comment]
@dignifiedquire, yes. I don't know what to do about the missing CLA from @srawlins.

File: lib/launcher.js [hunk 1]
```diff
@@ -11,7 +11,7 @@ var processDecorator = require('./launchers/process').decoratorFactory
 
 // TODO(vojta): remove once nobody uses it
 var baseBrowserDecoratorFactory = function (baseLauncherDecorator, captureTimeoutLauncherDecorator,
-  retryLauncherDecorator, processLauncherDecorator) {
+                                            retryLauncherDecorator, processLauncherDecorator) {
```
Comment (dignifiedquire) [inline lib/launcher.js]: Let's not do this please, instead use this indentation:

```js
var name = function (
  arg1,
  arg2,
  arg3
) {
  // body
}
```

File: test/e2e/load.feature [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      singleRun = false
+      """
+    When I start Karma
+    Then it fails with like:
+      """
+      Cannot load browser "NonExistingBrowser": it is not registered! Perhaps you are missing some plugin\?
+      """
+    And it fails with like:
+      """
+Found 1 load error
```
Comment (dignifiedquire) [inline test/e2e/load.feature]: Indentation?

File: lib/launcher.js [hunk 3]
```diff
@@ -20,7 +20,7 @@ var baseBrowserDecoratorFactory = function (baseLauncherDecorator, captureTimeou
   }
 }
 
-var Launcher = function (emitter, injector) {
+var Launcher = function (server, emitter, injector) {
```
Comment (dignifiedquire) [inline lib/launcher.js]: Are we sure this is not going to cause any issues in plugins?

Comment (budde377) [inline lib/launcher.js]: If we don't want to change the signature, I can just pull out the server from the injector directly, as I did in `lib/preprocessor.js`.

Comment (dignifiedquire) [inline lib/launcher.js]: Hmm not sure, what do you think?

**Comment by dignifiedquire** [comment]
Closes #855

**Comment by dignifiedquire** [comment]
Thanks a lot @budde377 and @srawlins for making this happen!

**Comment by budde377** [comment]
@dignifiedquire I just want to notice, that I didn't do the whole extract server from injector thing, so we might see some launchers breaking if they rely on the signature of the launcer.

**Comment by dignifiedquire** [comment]
@budde377 yeah I realised that, but I wanted to get it in, will fix reporters when they tell us, this is not really documented stable api anyway

**Pull request merged**